### ISSUE TITLE
ddcutil detect sometimes reports no DDC support (DELL P2715Q)

### ISSUE THREAD

**Issue by rombert** [rockowitz/ddcutil#58]
When running `ddcutil detect` I get back a negative response:

```
$ ddcutil detect
Display 1
   I2C bus:             /dev/i2c-10
   Supports DDC:        false
   EDID synopsis:
      Mfg id:           DEL
      Model:            DELL P2715Q
      Serial number:    V48W271KBPRL
      Manufacture year: 2017
      EDID version:     1.4
   VCP version:         2.1
```

However, the monitor supports DDC and it is enabled in the monitor menu. After running a simple operation, e.g. `ddcutil -d 1 setvcp 60 0xf` ( which basically sets the display to the already active one ), `ddcutil detect` now considers that the monitor supports DDC

```
$ ddcutil detect
Display 1
   I2C bus:             /dev/i2c-10
   Supports DDC:        true
   EDID synopsis:
      Mfg id:           DEL
      Model:            DELL P2715Q
      Serial number:    V48W271KBPRL
      Manufacture year: 2017
      EDID version:     1.4
   VCP version:         2.1

```

A second ddcutil detect call reverts to considering there is no DDC support though.

Version information:

```
$ ddcutil -V
ddcutil 0.9.1
Built without support for AMD Display Library (AMD proprietary driver).
Built with support for USB connected displays.
Built without function failure simulation.
Built with libdrm services.

Copyright (C) 2015-2018 Sanford Rockowitz
License GPLv2: GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by rockowitz** [comment]
Robert,

Interesting bug report.  Is that the full output of "ddcutil detect"?  
Please run "ddcutil detect --verbose --trace ddc" and send the output as 
an attachment.  Thank you.

Sanford

**Comment by rombert** [comment]
Hi Sanford,

Please see the attachments below. Note that the NVidia card is the one connected to the display. Thanks!


[ddcutil-interrogate.txt](https://github.com/rockowitz/ddcutil/files/2224858/ddcutil-interrogate.txt)
[ddcutil-detect.txt](https://github.com/rockowitz/ddcutil/files/2224822/ddcutil-detect.txt)

**Comment by rockowitz** [comment]
Robert,

This is baffling.   The code is executing as if DDC detected, but the 
output says it's not.   Let's try uping the tracing.

Please try:

   ddcutil detect --verbose --trace ddc  --trace i2c

Thanks,
Sanford

**Comment by rombert** [comment]
Hi Sanford,

The output is attached.

Robert

[ddcutil-more-trace.txt](https://github.com/rockowitz/ddcutil/files/2225605/ddcutil-more-trace.txt)

**Comment by rockowitz** [comment]
Robert,

I have uploaded a test build of ddcutil that contains additional tracing 
in the I2C layer.  It can be found at 
http://www.ddcutil.com/tarballs/ddcutil-0.9.2.tgz

If you're comfortable building a private copy of ddcutil, please 
download and build it.   Build instructions can be found at 
http://www.ddcutil.com/building

Then execute the following command and send the output:
   ddcutil detect --verbose --trace ddc --trace i2c


Also, if possible, install i2cdetect on your system, run the following 
command and send the output:
   i2cdetect -y 10

The above command assumes that the /dev/i2c-10 is still the relevant 
port.   If not, you can run the following command which scans all 
possible /dev/i2c devices:

   ddcutil environment --verbose

Thanks,
Sanford

**Comment by rockowitz** [comment]
Robert, 

You can hold off on using the test build I just posted.  The analysis has moved beyond the tracing added.

Sanford

**Comment by rockowitz** [comment]
Robert, 
Please go ahead and build the test version of ddcutil if possible.

Sanford

**Comment by rockowitz** [comment]
Robert,

VCP feature 60, which switches input, is a special case.   Some monitors 
will respond to feature requests no matter what the current input, 
others respond only on the current input.

If you do not issue a "setvcp 60" command, do the following commands 
work?  Do they cause "ddcutil detect" to change its "Supports DDC" report?
    ddcutil getvcp 60
    ddcutil getvcp 10
    ddcutil setvcp 10 + 1

Sanford

**Comment by rombert** [comment]
>  Please go ahead and build the test version of ddcutil if possible.

I get a 404 from that build, can you please re-check if it's there?

>  i2cdetect -y 10

[i2cdetect.txt](https://github.com/rockowitz/ddcutil/files/2228966/i2cdetect.txt)

>  ddcutil getvcp 60

   Supports DDC:        false

> ddcutil getvcp 10

   Supports DDC:        false

> ddcutil setvcp 10 + 1

   Supports DDC:        false

Thanks!

Robert

**Comment by rockowitz** [comment]
So i2cdetect does not detect I2c slave address x37 either.   Looks like 
this is a monitor-specific issue.

Latest tarball (I copied and pasted from my browser): 
http://www.ddcutil.com/tarballs/ddcutil-0.9.2.tar.gz

The latest version of ddcutil contains a slightly different test for the 
presence of slave address x37.  It works here, but then again so did the 
original test.  Will be interesting to see how your monitor responds.

Thanks again for you help in testing this.

Sanford

**Comment by rombert** [comment]
>  The latest version of ddcutil contains a slightly different test for the presence of slave address x37. It works here, but then again so did the original test. Will be interesting to see how your monitor responds.

Here's the output with the tarball you just linked above.

[ddcutil-detect-new-version.txt](https://github.com/rockowitz/ddcutil/files/2229212/ddcutil-detect-new-version.txt)

Thanks,

Robert

**Comment by rockowitz** [comment]
Robert,

As you've seen, the updated test for slave address works, and "Supports 
DDC" is reported as true.  Instead of attempting to read a single byte 
to the slave address, the test writes a single byte (x00) to the slave 
address.

The slave address x37 test is a bit vestigial.  Monitor detection now 
actually attempts to read features, which entail a write followed by a 
read, and the x37 test has become purely informational.   Which is why 
ddcutil can read and write features even though the x37 test fails, and 
ddcutil detect reports that DDC is unsupported.

Bottom line for you: ddcutil is working fine, just ignore the "Supports 
DDC: false" message.  (Or continue to use the test version you just built.)

Bottom line for me:   The x37/"Supports DDC" test needs to be reworked.  
I'm hesitant to make such a core change without extensive testing, as 
there may be monitors that fail the new test, and it's just a reporting 
problem. Will require some thought.

Thanks again for all your help in diagnosing this issue.

Regards,
Sanford

**Comment by rombert** [comment]
Hi Sanford,

> Bottom line for you: ddcutil is working fine, just ignore the "Supports DDC: false" message. (Or continue to use the test version you just built.)

That works for me, thanks!

Feel free to close this if you want, I'll live with the "false" message for now.

Thanks,

Robert

**Comment by rockowitz** [comment]
Hi Robert,

I'm going to leave the issue open until the final fix is in.  The edge 
cases have edge cases.  I may ask you to test another version.   Thanks 
again for your help.

Regards,
Sanford

**Comment by rockowitz** [comment]
Hi Robert,

The current test build of ddcutil 
(http://www.ddcutil.com/tarballs/ddcutil-0.9.2.tar.gz) has some revised 
code for testing slave address x37.   If you can, please build it, run 
"ddcutil detect", and send the output.   No need to enable tracing, the 
trace messages are already turned on in relevant functions.

Thanks,
Sanford

**Comment by rockowitz** [comment]
On 08/08/2018 05:24 AM, Robert Munteanu wrote:
>
> Hi Sandford,
>
> Sorry for the delayed response. For the moment I no longer have access 
> to that DELL P2715Q monitor. However, I will be able to test with a 
> DELL U2715H in a week. Hopefully that will trigger the same issues and 
> I will be able to confirm whether the new build helps.
>

If the new build fails with a Dell U2715H, that would be interesting.  
But if it succeeds, unless the display controller and firmware version 
match it won't mean much.   But it will be interesting to see.

**Comment by rombert** [comment]
For the record, I tested the old build with a U2175H, and it succeeds out-of-the-box. So I won't be able to provide updated information, sorry.

**Comment by tdakhran** [comment]
Faced a similar issue with DELL U2715H.
```shell
Invalid display
   I2C bus:  /dev/i2c-12
   DRM connector:           card0-DP-3
   EDID synopsis:
      Mfg id:               DEL - Dell Inc.
      Model:                DELL U2715H
      Product code:         53351  (0xd067)
      Serial number:        GH85D7B71G0S
      Binary serial number: 826749011 (0x31473053)
      Manufacture year:     2017,  Week: 45
   DDC communication failed
```
Tried different versions of `ddcutil` and Linux kernels.

In the end, a factory reset of the monitor and a power cycle of the monitor (unplug, wait, replug) helped.
```shell
   I2C bus:  /dev/i2c-12
   DRM connector:           card0-DP-3
   EDID synopsis:
      Mfg id:               DEL - Dell Inc.
      Model:                DELL U2715H
      Product code:         53351  (0xd067)
      Serial number:        GH85D7B71G0S
      Binary serial number: 826749011 (0x31473053)
      Manufacture year:     2017,  Week: 45
   VCP version:         2.1
```

**Comment by rockowitz** [comment]
@tdakhran That you for posting your solution. It's likely to be helpful to others.

That said, if the problem recurs please reach out.  The code that performs detection has changed and should be more tolerant of monitor peculiarities, and is more traceable for remote diagnosis.

By coincidence I just picked up a Dell U2719DX, which should be quite similar, so if you do identify issues I'll be in a better position to diagnose them. 

By chance, do you use or have you tried to use MST passthrough on this monitor?  With what video driver?  At least with amdgpu and i915 I expect the monitor will appear as two different /dev/i2c devices on two different DRM connectors.

**Comment by tdakhran** [comment]
> That said, if the problem recurs please reach out. The code that performs detection has changed and should be more tolerant of monitor peculiarities, and is more traceable for remote diagnosis.
Thank you for the awesome util and for keeping it up to date.

> By chance, do you use or have you tried to use MST passthrough on this monitor? With what video driver? At least with amdgpu and i915 I expect the monitor will appear as two different /dev/i2c devices on two different DRM connectors.

I don't know what MST passthrough is and probably I didn't use it. The laptop I have is Dell XPS 2022 and it doesn't have native video outputs. The monitor is connected via USB-C <-> HDMI adapter and only i915 is used for rendering via Wayland. Nvidia GPU is in a deep sleep state and is only used for CUDA.
